### PR TITLE
Implement NSObject Validation KVC methods

### PR DESCRIPTION
**Pull request by aballway**
Implements:
setNilValueForKey:
validateValue:forKey:error:
validateValue:forKeyPath:error:

Partially completes #2459, the Set and OrderedSet methods will be implemented in a separate CR because they are only tangentially related and will require a good bit of code.

### CODE REVIEW

**Comment by DHowett-MSFT** [comment]
nit: log when the validation method is stupid

File: Frameworks/Foundation/NSKeyValueCoding.mm [hunk 1]
```diff
@@ -457,6 +460,13 @@ bool KVCSetViaIvar(NSObject* self, struct objc_ivar* ivar, id value) {
 }
 
 /**
+ @Status Interoperable
+*/
+- (void)setNilValueForKey:(NSString*)key {
+    [NSException raise:NSInvalidArgumentException format:@"Attempt to set nil for key %@", key];
```
Comment (DHowett-MSFT) [inline Frameworks/Foundation/NSKeyValueCoding.mm]: nit: include `-[NSObject setValueForKey:]` in the message

File: Frameworks/Foundation/NSKeyValueCoding.mm [hunk 2]
```diff
@@ -578,30 +588,73 @@ - (NSMutableOrderedSet*)mutableOrderedSetValueForKeyPath:(NSString*)keyPath {
...
-*/
-- (void)setNilValueForKey:(NSString*)key {
-    UNIMPLEMENTED();
+static SEL KVCValidatorForPropertyName(NSObject* self, const char* key) {
+    size_t len = strlen(key);
+    // For the key "example", we must construct the following buffer:
+    // _ _ _ _ _ _ _ _ _ x a m p l e _ _ _ _ _ _ _ \0
+    // and fill it with the following characters:
+    // v a l i d a t e E x a m p l e : e r r o r : \0
+    std::unique_ptr<char[]> chars(new char[8 + len + 8]{});
```
Comment (DHowett-MSFT) [inline Frameworks/Foundation/NSKeyValueCoding.mm]: avoid a heap allocation by using alloca... or just a locally-sized array. no need for this heap nonsense :)

File: Frameworks/Foundation/NSKeyValueCoding.mm [hunk 3]
```diff
@@ -578,30 +588,73 @@ - (NSMutableOrderedSet*)mutableOrderedSetValueForKeyPath:(NSString*)keyPath {
...
+    chars[8 + len + 5] = 'r';
+    chars[8 + len + 6] = ':';
+    SEL sel = sel_getUid(chars.get());
+    return ([self respondsToSelector:sel]) ? sel : nullptr;
 }
 
 /**
- @Status Stub
- @Notes
+ @Status Caveat
```
Comment (DHowett-MSFT) [inline Frameworks/Foundation/NSKeyValueCoding.mm]: this is not a caveat, it is interoperable

File: Frameworks/Foundation/NSKeyValueCoding.mm [hunk 4]
```diff
@@ -578,30 +588,73 @@ - (NSMutableOrderedSet*)mutableOrderedSetValueForKeyPath:(NSString*)keyPath {
...
+ @Notes Invoked validate method must return BOOL
 */
 - (BOOL)validateValue:(id _Nullable*)ioValue forKey:(NSString*)key error:(NSError* _Nullable*)outError {
-    UNIMPLEMENTED();
-    return StubReturn();
+    SEL validationSelector = KVCValidatorForPropertyName(self, [key UTF8String]);
+    if (validationSelector) {
+        Method sig = class_getInstanceMethod([self class], validationSelector);
```
Comment (DHowett-MSFT) [inline Frameworks/Foundation/NSKeyValueCoding.mm]: we may want to prefer `[class methodSignatureForSelector:]`; this is a safe override point for subclassers.

Some of our other KVC code may run afoul of this.

File: Frameworks/Foundation/NSKeyValueCoding.mm [hunk 5]
```diff
@@ -578,30 +588,73 @@ - (NSMutableOrderedSet*)mutableOrderedSetValueForKeyPath:(NSString*)keyPath {
...
+        // 4 arguments: self, selector, ioValue, outError
+        if (sig && method_getNumberOfArguments(sig) == 4) {
```
Comment (DHowett-MSFT) [inline Frameworks/Foundation/NSKeyValueCoding.mm]: and then you can check the method signature's properties

File: Frameworks/Foundation/NSKeyValueCoding.mm [hunk 6]
```diff
@@ -578,30 +588,73 @@ - (NSMutableOrderedSet*)mutableOrderedSetValueForKeyPath:(NSString*)keyPath {
...
+            std::unique_ptr<char[]> returnType{ method_copyReturnType(sig) };
```
Comment (DHowett-MSFT) [inline Frameworks/Foundation/NSKeyValueCoding.mm]: without copying the return type

File: Frameworks/Foundation/NSKeyValueCoding.mm [hunk 7]
```diff
@@ -578,30 +588,73 @@ - (NSMutableOrderedSet*)mutableOrderedSetValueForKeyPath:(NSString*)keyPath {
...
+            // Reference platform does not check types, but will segfault when trying to return a value with size larger than BOOL
+            // And we will run into issues with values smaller than BOOL so simply ignore non-BOOL sized returns
+            if (objc_sizeof_type(returnType.get()) == sizeof(BOOL)) {
+                return ((BOOL(*)(id, SEL, id*, NSError**))objc_msgSend)(self, validationSelector, ioValue, outError);
```
Comment (DHowett-MSFT) [inline Frameworks/Foundation/NSKeyValueCoding.mm]: methodReturnLength~

Comment (DHowett-MSFT) [inline Frameworks/Foundation/NSKeyValueCoding.mm]: early return ;P

File: Frameworks/include/NSObject_NSKeyValueArrayAdapter-Internal.h [hunk 8]
```diff
@@ -14,16 +14,17 @@
...
-#include <Foundation/NSArray.h>
-#include <Foundation/NSMutableArray.h>
+#import <Foundation/NSMutableArray.h>
+#import <Foundation/NSMutableSet.h>
+#import <Foundation/NSMutableOrderedSet.h>
 
 @class _NSMutableKeyProxyArray;
 
-@interface _NSKeyProxyArray: NSArray
+@interface _NSKeyProxyArray : NSArray
```
Comment (DHowett-MSFT) [inline Frameworks/include/NSObject_NSKeyValueArrayAdapter-Internal.h]: untouched file, please revert?

File: include/Foundation/NSKeyValueCoding.h [hunk 9]
```diff
@@ -48,16 +48,16 @@ FOUNDATION_EXPORT NSString* const NSUnionOfSetsKeyValueOperator;
 - (id)valueForUndefinedKey:(NSString*)key;
 - (NSMutableArray*)mutableArrayValueForKey:(NSString*)key;
 - (NSMutableArray*)mutableArrayValueForKeyPath:(NSString*)keyPath;
-- (NSMutableSet*)mutableSetValueForKey:(NSString*)key STUB_METHOD;
-- (NSMutableSet*)mutableSetValueForKeyPath:(NSString*)keyPath STUB_METHOD;
-- (NSMutableOrderedSet*)mutableOrderedSetValueForKey:(NSString*)key STUB_METHOD;
-- (NSMutableOrderedSet*)mutableOrderedSetValueForKeyPath:(NSString*)keyPath STUB_METHOD;
+- (NSMutableSet*)mutableSetValueForKey:(NSString*)key;
```
Comment (DHowett-MSFT) [inline include/Foundation/NSKeyValueCoding.h]: hey hey hey now let's not get ahead of ourselves

File: tests/UnitTests/Foundation/NSObject_NSKeyValueValidation.mm [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//******************************************************************************
+
+#include <Foundation/Foundation.h>
+#include <TestFramework.h>
+
+struct smolStruct {
```
Comment (DHowett-MSFT) [inline tests/UnitTests/Foundation/NSObject_NSKeyValueValidation.mm]: aaaaaaaaaaaaaaaaaaaaaaaaa y

File: tests/UnitTests/Foundation/NSObject_NSKeyValueValidation.mm [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+    EXPECT_EQ(NO, [test validateValue:nil forKeyPath:@"obj.obj" error:&error]);
+    EXPECT_EQ(YES, [test validateValue:&data forKeyPath:@"obj.obj" error:&error]);
+    EXPECT_OBJCEQ([NSData data], data);
+}
+
+TEST(NSObject, ValidateValueForKeyPathShouldProperlyFollowKeyPath) {
+    ValidationHolder* test = [ValidationHolder holder];
+    NSData* data = [NSData data];
+    NSError* error = nil;
+    EXPECT_EQ(NO, [test validateValue:&data forKeyPath:@"obj.valid1" error:&error]);
```
Comment (DHowett-MSFT) [inline tests/UnitTests/Foundation/NSObject_NSKeyValueValidation.mm]: do you know if the validators are ever called automatically by the value setters?

File: tests/UnitTests/Foundation/NSObject_NSKeyValueValidation.mm [hunk 12]
```diff
@@ -0,0 +1,248 @@
...
+
+    EXPECT_ANY_THROW([test validateValue:&data forKeyPath:@"nonexistent.valid1" error:&error]);
+
+    EXPECT_EQ(YES, [test validateValue:&data forKeyPath:@"obj.invalid1" error:&error]);
+    EXPECT_EQ(YES, [test validateValue:&data forKeyPath:@"obj.nonexistent" error:&error]);
+}
+
+TEST(NSObject, SetNilValueForKey) {
+    ValidationHolder* test = [ValidationHolder holder];
+    EXPECT_ANY_THROW([test setNilValueForKey:@"obj"]);
```
Comment (DHowett-MSFT) [inline tests/UnitTests/Foundation/NSObject_NSKeyValueValidation.mm]: we must also test the consumption of this method by `setValue:nil forKey:x`

File: tests/UnitTests/Foundation/NSObject_NSKeyValueValidation.mm (hunk 10 above)
Comment (aballway) [inline tests/UnitTests/Foundation/NSObject_NSKeyValueValidation.mm]: It is very teeny

File: include/Foundation/NSKeyValueCoding.h (hunk 9 above)
Comment (aballway) [inline include/Foundation/NSKeyValueCoding.h]: Whoops left this from when I arrogantly thought I could fit all of these changes in one PR 😈

File: tests/UnitTests/Foundation/NSObject_NSKeyValueValidation.mm (hunk 11 above)
Comment (aballway) [inline tests/UnitTests/Foundation/NSObject_NSKeyValueValidation.mm]: I wasn't able to get it called by value setters, and documentation seems to suggest it's something that consumers use manually, so I don't believe so

File: Frameworks/Foundation/NSKeyValueCoding.mm [hunk 13]
```diff
@@ -477,6 +486,11 @@ - (void)setValue:(id)val forKey:(NSString*)key {
     }
 
     const char* rawKey = [key UTF8String];
+
+    if (!val) {
+        switch (rawKey[0]) { OBJC_APPLY_NUMERIC_TYPE_ENCODINGS(FAIL_IF_SCALAR); }
```
Comment (DHowett-MSFT) [inline Frameworks/Foundation/NSKeyValueCoding.mm]: I don't get it. Can't you just call `[self setNilValueForKey:]`? Or is this to get the appropriate "is not key-value coding compliant" exception?

File: Frameworks/Foundation/NSKeyValueCoding.mm [hunk 14]
```diff
@@ -612,25 +626,32 @@ static SEL KVCValidatorForPropertyName(NSObject* self, const char* key) {
...
 */
 - (BOOL)validateValue:(id _Nullable*)ioValue forKey:(NSString*)key error:(NSError* _Nullable*)outError {
     SEL validationSelector = KVCValidatorForPropertyName(self, [key UTF8String]);
     if (validationSelector) {
-        Method sig = class_getInstanceMethod([self class], validationSelector);
+        NSMethodSignature* sig = [self methodSignatureForSelector:validationSelector];
         // 4 arguments: self, selector, ioValue, outError
-        if (sig && method_getNumberOfArguments(sig) == 4) {
-            std::unique_ptr<char[]> returnType{ method_copyReturnType(sig) };
+        if (sig && sig.numberOfArguments == 4) {
```
Comment (DHowett-MSFT) [inline Frameworks/Foundation/NSKeyValueCoding.mm]: nit: nil checking sig is not required, since `nil.numberOfArguments` == `0`

File: Frameworks/Foundation/NSKeyValueCoding.mm (hunk 13 above)
Comment (aballway) [inline Frameworks/Foundation/NSKeyValueCoding.mm]: no but I do need to check the type encoding rather than the key...

File: Frameworks/Foundation/NSKeyValueCoding.mm [hunk 15]
```diff
@@ -405,12 +405,16 @@ static bool quickSet(id self, SEL setter, id value, const char* valueType) {
 
 #define QUICK_SET_CASE(type, name, capitalizedName, encodingChar) \
     case encodingChar:                                            \
+        if (!value) {                                             \
```
Comment (DHowett-MSFT) [inline Frameworks/Foundation/NSKeyValueCoding.mm]: this one _will_ actually throw for `id`; is that expected?

File: Frameworks/Foundation/NSKeyValueCoding.mm [hunk 16]
```diff
@@ -425,6 +421,11 @@ bool KVCSetViaAccessor(NSObject* self, SEL setter, id value, NSString* key) {
     if (sig && [sig numberOfArguments] == 3) {
         const char* valueType = [sig getArgumentTypeAtIndex:2];
 
+        if (!value && !(valueType[0] == '@' || valueType[0] == '#')) {
```
Comment (DHowett-MSFT) [inline Frameworks/Foundation/NSKeyValueCoding.mm]: so `^` will happily splode?

File: Frameworks/Foundation/NSKeyValueCoding.mm [hunk 17]
```diff
@@ -578,30 +602,80 @@ - (NSMutableOrderedSet*)mutableOrderedSetValueForKeyPath:(NSString*)keyPath {
...
+        NSMethodSignature* sig = [self methodSignatureForSelector:validationSelector];
+        // 4 arguments: self, selector, ioValue, outError
+        if (sig.numberOfArguments == 4) {
+            // Reference platform does not check types, but will segfault when trying to return a value with size larger than BOOL
+            // And we will run into issues with values smaller than BOOL so simply ignore non-BOOL sized returns
+            if (sig.methodReturnLength == sizeof(BOOL)) {
+                return ((BOOL(*)(id, SEL, id*, NSError**))objc_msgSend)(self, validationSelector, ioValue, outError);
+            } else {
+                // But give a warning because this is stricter than the reference platform
+                TraceWarning(TAG,
```
Comment (DHowett-MSFT) [inline Frameworks/Foundation/NSKeyValueCoding.mm]: you could also stuff an error into outError (!) if you wanted!

File: Frameworks/Foundation/NSKeyValueCoding.mm (hunk 16 above)
Comment (DHowett-MSFT) [inline Frameworks/Foundation/NSKeyValueCoding.mm]: AJ answered offline: yes.

File: Frameworks/include/NSObject_NSKeyValueCoding-Internal.h [hunk 18]
```diff
@@ -24,5 +24,5 @@ SEL KVCGetterForPropertyName(NSObject* self, const char* key);
 bool KVCGetViaAccessor(NSObject* self, SEL getter, id* ret);
 bool KVCGetViaIvar(id self, struct objc_ivar* ivar, id* ret);
 SEL KVCSetterForPropertyName(NSObject* self, const char* key);
-bool KVCSetViaAccessor(NSObject* self, SEL setter, id value);
-bool KVCSetViaIvar(NSObject* self, struct objc_ivar* ivar, id value);
\ No newline at end of file
+bool KVCSetViaAccessor(NSObject* self, SEL setter, id value, NSString* key = nil);
+bool KVCSetViaIvar(NSObject* self, struct objc_ivar* ivar, id value, NSString* key = nil);
```
Comment (rajsesh-msft) [inline Frameworks/include/NSObject_NSKeyValueCoding-Internal.h]: nit: make git happy and insert new line at end.

File: tests/UnitTests/Foundation/NSObject_NSKeyValueValidation.mm [hunk 19]
```diff
@@ -0,0 +1,261 @@
...
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//******************************************************************************
+
+#include <Foundation/Foundation.h>
+#include <TestFramework.h>
+
+struct smolStruct {
+    char a, b;
```
Comment (rajsesh-msft) [inline tests/UnitTests/Foundation/NSObject_NSKeyValueValidation.mm]: coding conventions dictate 
`struct smolStruct {
    char a;
    char b;
};
`

File: include/Foundation/NSKeyValueCoding.h (hunk 9 above)
Comment (rajsesh-msft) [inline include/Foundation/NSKeyValueCoding.h]: Why not? You've been known to do that before :)

---
In reply to: [113270799](https://github.com/Microsoft/WinObjC/pull/2610#discussion_r113270799) [](ancestors = 113270799)

File: Frameworks/Foundation/NSKeyValueCoding.mm (hunk 17 above)
Comment (aballway) [inline Frameworks/Foundation/NSKeyValueCoding.mm]: I'm not sure that would fit the contract of the method.  As for testing validation the NSObject_ValidateValueForKeyPath::Test validates the validation method being called
